OpenRA.Game/Graphics/WorldRenderer.cs

```diff
@@ -66,11 +67,24 @@ internal WorldRenderer(ModData modData, World world)
 			palette.Initialize();
 
 			TerrainLighting = world.WorldActor.TraitOrDefault<ITerrainLighting>();
+			renderers = world.WorldActor.TraitsImplementing<IRenderer>().ToArray();
 			terrainRenderer = world.WorldActor.TraitOrDefault<IRenderTerrain>();
 
 			debugVis = Exts.Lazy(() => world.WorldActor.TraitOrDefault<DebugVisualizations>());
 		}
 
+		public void BeginFrame()
+		{
+			foreach (var r in renderers)
+				r.BeginFrame();
+		}
+
+		public void EndFrame()
+		{
+			foreach (var r in renderers)
+				r.EndFrame();
```

Comment:
Questions that come to mind.

Renderers would be independant of each other? No cross refences or shared state? Would some IRenderables references multiple renderers?

I see you pass the renderable when creating the IRenderable.

So this would mean the renderes could be active/called in any order. 

This only ensures the order in which begin and end frame is called.

Might be an option to put MapEditor DebugRenderables in seperate renderer. But may not have much added value.

Reply:
This will get much more polished once we turn SpriteRenderers into a trait. This current solution is as minimal as possible